Update readme with beekeeper-api information (#103)
ExpediaGroup/beekeeper: README.md

![Logo](.README_images/full_logo.png)

![Java CI](https://github.com/ExpediaGroup/beekeeper/workflows/Java%20CI/badge.svg?event=push) [![Coverage Status](https://coveralls.io/repos/github/ExpediaGroup/beekeeper/badge.svg?branch=main)](https://coveralls.io/github/ExpediaGroup/beekeeper?branch=main) ![GitHub license](https://img.shields.io/github/license/ExpediaGroup/beekeeper.svg)
# Table of contents 

> - [Overview](#overview)

Reviewer: Minor. Headings should be written using Title Case.

Beekeeper makes use of [Apiary](https://github.com/ExpediaGroup/apiary) - an open source federated cloud data lake - to detect changes in the Hive Metastore. One of Apiary’s components, the [Apiary Metastore Listener](https://github.com/ExpediaGroup/apiary-extensions/tree/main/apiary-metastore-events/sns-metastore-events/apiary-metastore-listener), captures Hive events and publishes these as messages to an SNS topic. Beekeeper uses these messages to detect changes to the Hive Metastore, and perform appropriate deletions.

> Beekeeper is comprised of four separate Spring-based Java applications:

Reviewer 2: Might be helpful to include a contents table with hyperlinks at the top of this README with all the titles of the sections, that way you can quickly see whats included and jump straight there

1. Scheduler Apiary - An application that schedules paths and metadata for deletion in a shared database, with one table for unreferenced paths and another for expired metadata. 
2. Path Cleanup - An application that perform deletions of unreferenced paths.
3. Metadata Cleanup - An application that perform deletions of expired metadata.

> 4. Beekeeper API - An application that allows users to see what metadata and paths are in the database.

Reviewer 3: An application = A Rest API that....

Beekeeper makes use of [Apiary](https://github.com/ExpediaGroup/apiary) - an open source federated cloud data lake - to detect changes in the Hive Metastore. One of Apiary’s components, the [Apiary Metastore Listener](https://github.com/ExpediaGroup/apiary-extensions/tree/main/apiary-metastore-events/sns-metastore-events/apiary-metastore-listener), captures Hive events and publishes these as messages to an SNS topic. Beekeeper uses these messages to detect changes to the Hive Metastore, and perform appropriate deletions.

Beekeeper is comprised of four separate Spring-based Java applications:
1. Scheduler Apiary - An application that schedules paths and metadata for deletion in a shared database, with one table for unreferenced paths and another for expired metadata. 
2. Path Cleanup - An application that perform deletions of unreferenced paths.
3. Metadata Cleanup - An application that perform deletions of expired metadata.

> 4. Beekeeper API - A REST API that allows users to see what metadata and paths are in the database.

Reviewer: I think we should drop `users` from this and all the other references. An API is meant to be used in an ad-hoc manner by people or other applications.

# Running Beekeeper

> Beekeeper consists of four Spring Boot applications which run independently of each other:

Reviewer: This could get simplified by saying that all Beekeeper modules are run independently (without mentioning how many are available). Mainly, to avoid having to update the number every time there's a new module.

For example, the healthcheck endpoint at: `http://<address>:<port>/actuator/health`. 

> By default, `beekeeper-scheduler-apiary` listens on port 8080, `beekeeper-path-cleanup` listens on port 8008, `beekeeper-metadata-cleanup` listens on 9008 and `beekeeper-api` listens on 7008. To access this endpoint when running in a Docker container, the port must be published:

Reviewer 2: Might be nice to put this in a table?

| `scheduler-delay-ms`                | No       | Amount of time (in milliseconds) between consecutive cleanups. Default value is `300000` (5 minutes after the previous cleanup completes). |
| `Metastore-uri`                     | Yes      | URI of the Hive Metastore where tables to be cleaned-up are located. |

> ## Beekeeper-API

Reviewer 3: can you mention the swagger enpoint somewhere in here and how to reach it when beekeper api is deployed. It's a good starting point for most users to explore capabiities

> Beekeeper also has an API which provides access to the Beekeeper database and allows to see what metadata and paths are currently being managed.

Reviewer 2: ```suggestion
Beekeeper also has an API which provides read access to the Beekeeper database and allows you to see what metadata and paths are currently being managed.
```

> This allows the user to manually enter a database and a table name and check whether this table has been successfully registered in Beekeeper along with the TTL configuration: the current status of the table, the date and time it will be deleted, the current cleanup delay... etc.

Reviewer 2: ```suggestion
The user can manually enter a database and a table name and check whether this table has been successfully registered in Beekeeper along with the TTL configuration, the current status of the table, the date and time it will be deleted, the current cleanup delay... etc.
```

It currently supports two endpoints; one for the expired metadata and another one for the unreferenced paths.

It also supports different filters (see [filtering section](#filtering)).

### Expired metadata endpoint (`GET /metadata`)

This endpoint will return the TTL configuration of all expired partitions that are going to be deleted (or have been deleted) in a specific table. If it is unpartitioned it will just show one object; the table.

It is available in this url; 

>     http://beekeeper-api.<address>/api/v1/database/{databaseName}/table/{tableName}/metadata

Reviewer 2: whats the address part here?

Author: it's `egdl-eks-us-east-1.egdp-prod.aws.away.black`; I wanted to put it in but it's open source and I don't know if that's giving private information? does that make sense?

 where `{databaseName}` and `{tableName}` must be replaced by the database and table name the user wants to search for. So for example, if they wanted to check a table called `my_cool_table` in the database `my_cool_database`, they would go to 
 

>     http://beekeeper-api.<address>/api/v1/database/my_cool_database/table/my_cool_table/metadata

Reviewer 2: probably best to mention the base url once above `http://beekeeper-api.<address>/api/v1/` and then in these sections specify the rest of the url, like `<base>/api/v1/database/my_cool_database/table/my_cool_table/metadata` that way if the base needs updating in the future (e.g. v2) it can be done in one place

### Unreferenced paths endpoint (`GET /unreferenced-paths`)

This endpoint will return the configuration of all unreferenced paths that are going to be deleted (or have been deleted) in a specific table. If it is unpartitioned it will just show one object; the table.

It is available in this url; 

    http://beekeeper-api.<address>/api/v1/database/{databaseName}/table/{tableName}/unreferenced-paths

>  where `{databaseName}` and `{tableName}` must be replaced by the database and table name the user wants to search for. So for example, if they wanted to check a table called `my_cool_table` in the database `my_cool_database`, they would go to 

Reviewer 2: so cool

Author: cool people, cool things

 
    http://beekeeper-api.<address>/api/v1/database/my_cool_database/table/my_cool_table/unreferenced-paths

### Filtering

The filtering available is the same in both endpoints, just keep in mind that in the first one it is referring to expired metadata and in the second one it is referring to unreferenced paths. 

The following table gives an overview of the filters available. It uses the metadata endpoint for the examples, but if the user wants to refer to paths they just have to replace `/database/{databaseName}/table/{tableName}/metadata` with `/database/{databaseName}/table/{tableName}/unreferenced-paths`.

> ![Filtering table](.README_images/filtering_table.png)

Reviewer 2: The screenshot is a little blurry. Would it be super long for you to copy the info into an actual table?
Also easier for people to use it as they can copy-and-paste and make less errors.

you could also add a column for 'available in both endpoints' and then the note at the bottom can be included in the table

| `scheduler-delay-ms`                | No       | Amount of time (in milliseconds) between consecutive cleanups. Default value is `300000` (5 minutes after the previous cleanup completes). |
| `Metastore-uri`                     | Yes      | URI of the Hive Metastore where tables to be cleaned-up are located. |

## Beekeeper-API

Beekeeper also has an API which provides read access to the Beekeeper database and allows you to see what metadata and paths are currently being managed.

> The user can manually enter a database and a table name and check whether this table has been successfully registered in Beekeeper along with things like the current status of the table, the date and time it will be deleted, the current cleanup delay... etc.

Reviewer: current cleanup delay, etc. (drop the ellipsis).

    http://<host>/swagger-ui.html

For the two main endpoints, the base url (will be referred to as `<base-url>` in the following sections) is

    http://<host>/api/v1

and the rest of the url will depend on which endpoint you want to access (see [Expired metadata endpoint](#expired-metadata-endpoint) and [Unreferenced paths endpoint](#unreferenced-paths-endpoint)).

It also supports different filters (see [filtering section](#filtering)).

### Expired metadata endpoint (`GET /metadata`)

This endpoint will return the TTL configuration of all expired partitions that are going to be deleted (or have been deleted) in a specific table. If it is unpartitioned it will just show one object; the table.

    <base-url>/database/{databaseName}/table/{tableName}/metadata

 where `{databaseName}` and `{tableName}` must be replaced by the database and table name the user wants to search for. So for example, if they wanted to check a table called `my_cool_table` in the database `my_cool_database`, they would go to 
 
    <base-url>/database/my_cool_database/table/my_cool_table/metadata
    

> The API will display all the partitions in that table unless it is unpartitioned, in that case it will show only one object; the table. If you want to check only the table object without all of its individual partitions, you'll have to search for the one with the variable `"partitionName"=null` as such:

Reviewer: You should change the second person style in this writing and adopt the third person to make it look more generic and professional; unless the entire document is written in the second person in which case is OK to keep it consistent.
Example: If you want...

>     "path": "s3://mybucket/mydatabase/mytable",

Reviewer: You should include the opening and closing curly brace to make it clear that it's an object.

    "databaseName": "mydatabase",
    "tableName": "mytable",
    "partitionName": null,
    "housekeepingStatus": "DELETED",
    "creationTimestamp": "2020-09-14T17:22:55",
    "modifiedTimestamp": "2020-09-14T18:36:52",
    "cleanupTimestamp": "2020-09-14T17:36:32",
    "cleanupDelay": "PT10M",
    "cleanupAttempts": 1,
    "lifecycleType": "EXPIRED"

> You can do this by using [filters](#filtering). If you want to search for the table object, you'll have to add a filter with the path to the table, for example 

Reviewer: Another example here of the second person being used.

    <base-url>/database/{databaseName}/table/{tableName}/metadata?path=s3://mybucket/mydatabase/mytable

### Unreferenced paths endpoint (`GET /unreferenced-paths`)

This endpoint will return the configuration of all unreferenced paths that are going to be deleted (or have been deleted) in a specific table. If it is unpartitioned it will just show one object; the table.

> The url is the same one as in the [metadata endpoint](#expired-metadata-endpoint), except instead of ending in `metadata` it ends in `unreferenced-paths`.

Reviewer: We could probably drop this. It's redundant. The path is mentioned in the heading and you have the full example in the section below.

It is available in this url; 

    <base-url>/database/{databaseName}/table/{tableName}/unreferenced-paths

 where `{databaseName}` and `{tableName}` must be replaced by the database and table name the user wants to search for. So for example, if they wanted to check a table called `my_cool_table` in the database `my_cool_database`, they would go to 
 
    <base-url>/database/my_cool_database/table/my_cool_table/unreferenced-paths

### Filtering

> The filtering available is the same in both endpoints, just keep in mind that in the first one it is referring to expired metadata and in the second one it is referring to unreferenced paths. 

Reviewer: This can also be dropped. It's clear that both endpoints are different but they share the same filtering capabilities.

> The following table gives an overview of the filters available. It uses the metadata endpoint for the examples, but if the user wants to refer to paths they just have to replace `/database/{databaseName}/table/{tableName}/metadata` with `/database/{databaseName}/table/{tableName}/unreferenced-paths`.

Reviewer: The section on replacing one with the other can be dropped. Just make it clear that both endpoints support the same filtering capabilities.
